### DISCUSSION TITLE
OpenTelemetry

### DISCUSSION THREAD

**Discussion by purkhusid** [vercel/next.js#32211] [Help]
Hello! I'm using NextJS to host a server side rendered website with a GraphQL endpoint that the client talks to.

I'm trying to implement some tracing for the server side and I'm hitting an issue because I need to make sure that the
first thing that is done when the application is loaded is that the tracing is initialized.
This is needed because the OpenTelemetry SDK hooks into the NodeJS runtime and libraries to enable some automated tracing.

Is there some way to configure NextJS so that I can inject an import/require statement at the very beginning of the bundle?

**Comment by weyert** [comment]
I am using a custom server for this at the moment. Easiest way to do this.  I still haven't solved the issue of resolving the route / path of the url, though so the cardinality (high) when using Opentelemetry Metrics SDK with the Prometheus exporter is problematic.

I was told during the announcement of Next.js 12 that Vercel is working on built-in tracing and metrics support into Next.js by @timneutkens but haven't much more about it yet.

**Reply by weyert** [reply]
I will ask work if I can make an example out of it, though

**Reply by purkhusid** [reply]
@weyert Yeah, if you have a custom server you can work around this but if you run the app on Vercel/Netlify or similar platforms you are out of luck. I'm currently running the site on Netlify and managed to find a workaround that initializes the tracer before anything else. The workaround is built around knowing that Netlify uses AWS lambda and there is a special environment variables in AWS lambda that allows you to wrap the function with a custom wrapper: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-modify.html

I'm hoping that the NextJS authors don't build their own custom tracing solution and decide to use OpenTelemetry since it's the de facto standard for tracing.

**Reply by weyert** [reply]
Ah yeah, I am running it in GKE.  Sounds like you are using `node --require tracing.js` or in your case `NODE_OPTIONS=--require tracing.js` approach then :)

In the past they use Opentelemetry for their internally tracing so I can imagine they would use it for tracing. Definitely agree with standard command. I am looking forward to the implementation of Metrics spec in Node.js

**Reply by afreakk** [reply]
@purkhusid could you elaborate more on your workaround?
Do you put the `NODE_OPTIONS` environment variable in `https://vercel.com/<ORG>/<PROJ>/settings/environment-variables` ?
I did that, and it seems to influence the front-end build also, so its freaking out because frontend cant find the tracing.js in project root.
```
Error: Cannot find module 'tracing.js'
--
13:52:32.869 | Require stack:
13:52:32.869 | - internal/preload
13:52:32.870 | at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
13:52:32.870 | at Function.Module._load (internal/modules/cjs/loader.js:746:27)
13:52:32.870 | at Module.require (internal/modules/cjs/loader.js:974:19)
13:52:32.870 | at Module._preloadModules (internal/modules/cjs/loader.js:1244:12)
13:52:32.870 | at loadPreloadModules (internal/bootstrap/pre_execution.js:475:5)
13:52:32.870 | at prepareMainThreadExecution (internal/bootstrap/pre_execution.js:72:3)
13:52:32.870 | at internal/main/run_main_module.js:7:1 {
13:52:32.870 | code: 'MODULE_NOT_FOUND',
13:52:32.870 | requireStack: [ 'internal/preload' ]
13:52:32.870 | }
13:52:32.871 | Sandbox exited with code 1 and signal null
```
Ah, I see you did it in netlify, not vercel, but should be pretty similar?

**Comment by purkhusid** [comment]
@afreakk Our hack requires that the platform uses AWS lambda behind the scenes. I'm not sure if Vercel is on AWS or if they are using GCP. AWS Lambda supports a special environment variable to wrap the executing binary using the `AWS_LAMBDA_EXEC_WRAPPER` environment variable.

I include my wrapper in the output of the Netlify build and it gets put into the `/var/task` folder inside the lambda so I set the `AWS_LAMBDA_EXEC_WRAPPER` to `/var/task/lambda_wrapper.sh`.

The contents of the wrapper is then:

```
#!/bin/bash

if [ -f ./buildtime_envs ]; then
    source ./buildtime_envs
else
    echo "No buildtime_envs file found"
fi

export NODE_OPTIONS="${NODE_OPTIONS} --require ./tracing.js"

if [[ $OTEL_RESOURCE_ATTRIBUTES != *"service.name="* ]]; then
  export OTEL_RESOURCE_ATTRIBUTES="service.name=service-name,faas.handler=${_HANDLER},faas.name=${AWS_LAMBDA_FUNCTION_NAME},faas.memory_size=${AWS_LAMBDA_FUNCTION_MEMORY_SIZE},faas.execution_env=${AWS_EXECUTION_ENV},cloud.provider=aws,cloud.region=${AWS_REGION},netlify.site_name=${SITE_NAME},${OTEL_RESOURCE_ATTRIBUTES},netlify.site_id=${SITE_ID},netlify.build_id=${BUILD_ID},netlify.context=${CONTEXT},git.repository=${REPOSITORY_URL},git.branch=${BRANCH},git.ref=${COMMIT_REF},netlify.deploy_url=${DEPLOY_URL},netlify.deploy_prime_url=${DEPLOY_PRIME_URL}"
fi

exec "$@"

```
Hopefully you can use this to do something similar with Vercel!

**Reply by afreakk** [reply]
Thanks for the inside info!
Seems to work, 
Added `AWS_LAMBDA_EXEC_WRAPPER` env variable pointing to `./serverlessFiles/lambda_wrapper.sh` which adds `--require ./serverlessFiles/serverlessTracing.js` to `NODE_OPTIONS`, which is a bundled javascript file.  
then in `vercel.json` add `"includeFiles": "{serverlessFiles,node_modules/jaeger-client}/**"` to the functions.
like so for insance:
```
    "functions": {
        "api/*.+(js|ts)": {
            "includeFiles": "{serverlessFiles,node_modules/jaeger-client}/**"
        }
    }
```   
So the files get shipped to aws-lambda.

For som reason some tracing module wanted to readFileSync a file inside node_modules/jaeger-client, so had to add that also to includeFiles.  
Can also bundle this `serverlessTracing.js` file during vercel build, by adding [vercel-build](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/node-js#custom-build-step-for-node.js) to package.json.
The reason you have to bundle, or do some other workaround, is cause vercel doesnt ship any node modules not explicitly imported by the function to aws-lambda, so it doesnt care about the dependencies of the file being --required in `NODE_OPTIONS`.

(This is for vercel vanilla using parcel build, not next.js, but I think its pretty similar)
